cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,403 @@
+        for i in range(j - 1, -1, -1):
+            if cnots[i][control] != cnots[j][control]:
+                break
+            # We take a note of the control (resp. target).
+            common_a[cnots[i][target]] = i
+
+        # Next, we look forward for as long as the controls (resp. targets) are the
+        # same. They all commute, so all are potential candidates for being simplified.
+        common_b: Dict[int, int] = {}
+        for k in range(j + 1, len(cnots)):
+            if cnots[j][target] != cnots[k][target]:
+                break
+            # We take a note of the target (resp. control).
+            common_b[cnots[k][control]] = k
+
+        # Among all the candidates, find if they have a match.
+        keys = common_a.keys() & common_b.keys()
+        for key in keys:
+            assert common_a[key] != common_b[key]
+            # We perform the swap which removes the pivot.
+            new_idx: List[int] = (
+                [idx for idx in range(0, j) if idx != common_a[key]]
+                + [common_b[key], common_a[key]]
+                + [idx for idx in range(j + 1, len(cnots)) if idx != common_b[key]]
+            )
+            # Since we removed the pivot, the length should be one fewer.
+            assert len(new_idx) == len(cnots) - 1
+            cnots = [cnots[idx] for idx in new_idx]
+            return True, cnots
+
+    return False, cnots
+
+
+def _simplify_cnots(cnots: List[Tuple[int, int]]) -> List[Tuple[int, int]]:
+    """Takes a series of CNOTs and tries to applies rule to cancel out gates.
+
+    Args:
+        cnots: A list of CNOTs represented as tuples of integer (control, target).
+
+    Returns:
+        A Boolean saying whether a simplification has been found.
+        The simplified list of CNOTs.
+    """
+
+    found_simplification = True
+    while found_simplification:
+        for simplify_fn, flip_control_and_target in itertools.product(
+            [_simplify_commuting_cnots, _simplify_cnots_triplets], [False, True]
+        ):
+            found_simplification, cnots = simplify_fn(cnots, flip_control_and_target)
+            if found_simplification:
+                break
+
+    return cnots
+
+
+def _get_gates_from_hamiltonians(
+    hamiltonian_polynomial_list: List['cirq.PauliSum'],
+    qubit_map: Dict[str, 'cirq.Qid'],
+    theta: float,
+    ladder_target: bool = False,
+) -> Generator['cirq.Operation', None, None]:
+    """Builds a circuit according to [1].
+
+    Args:
+        hamiltonian_polynomial_list: the list of Hamiltonians, typically built by calling
+            _build_hamiltonian_from_boolean().
+        qubit_map: map of string (boolean variable name) to qubit.
+        theta: A single float scaling the rotations.
+        ladder_target: Whether to use convention of figure 7a or 7b of [4]. The two formulations
+            yield the same output, but can be simplified differently.
+            For example for 3 qubits, the ladder target would be:
```

Comment:
```suggestion
            For example for 3 qubits, the `ladder_target=True` results in:
```

Reply:
Is my suggestion right? Wouldn't the other one look more like a ladder? :)

Reply:
Also, can we move this explanation to the public constructor of `BooleanHamiltonian`? Then, we can just refer to it from here, the private method that: `ladder_target: has the same meaning as on BooleanHamiltonian`.

Reply:
Yes, you are right, the other one looked more like a ladder. The code was correct, but the two examples in the comment were inverted, so I flipped them.

Moved to the constructor and changed the other comments.